src/BlockNLPModels.jl

```diff
@@ -0,0 +1,79 @@
+module BlockNLPModels
+
+using NLPModels
+using LinearAlgebra
+
+export AbstractBlockNLP
+export BlockNLPModel
+export DualizedNLPblockModel
+export FullSpaceModel
+
+abstract type AbstractBlockNLP end
+
+mutable struct BlockNLPModel{T<:AbstractNLPModel} <: AbstractBlockNLP
+  blocks::Vector{T}
+  linkconstraints::Vector{Float64}
+end
+
+mutable struct DualizedNLPblockModel{T, S} <: AbstractNLPModel{T, S}
+  meta::NLPModelMeta{T, S}
+  counters::Counters
+  problem_block::AbstractNLPModel
+  λ::AbstractVector # dual variables
+  a::AbstractVector # linking constraint coeff
+end
+
+function DualizedNLPblockModel(nlp::AbstractNLPModel, λ::AbstractVector, a::AbstractVector)
+  meta = nlp.meta
+  return DualizedNLPblockModel(meta, Counters(), nlp, λ, a)
+end
+
+function NLPModels.obj(nlp::DualizedNLPblockModel, x::AbstractVector)
+  n = nlp.problem_block.meta.nvar
+  d(x) = dot(nlp.λ, dot(nlp.a, x))
+  return obj(nlp.problem_block, x) + d(x)
+end
```

Comment:
Do we expect this to be a callback that will be implemented by the particular decomposition algorithm?

Reply:
Yes, everything that has been implemented for the data structure DualizedNLPblockModel is meant for the dual decomposition algorithm. You can take a look at the dual_decomposition.jl file in the test folder for more details. 

For ADMM, we will have struct AugmentedNLPblockModel.